### PR TITLE
Fix logic flaws in caching system (was Fix 457)

### LINKED ISSUE #457: Caching is disabled with default args
I expect a call to `cache` with default arguments to cache the result.  It does not.

The following code to cache does not cache:
```python
from nilearn._utils.cache_mixin import cache

def my_fun(X, Y)
     return X + Y
Z = cache(my_fun, memory='nilearn_cache')(1, 1)
```

Instead, the call must be:
```python
Z = cache(my_fun, memory='nilearn_cache', func_memory_level=1, memory_level=1)(1, 1)
```

I suggest tweaking the semantics of `func_memory_level`, such that if `func_memory_level is None`, everything gets cached.  Put it this way: if you call `cache` without specifying a `func_memory_level`, you always want it cached--rather, than *never* cached.

Alternately, I would like the function to raise a warning if `func_memory_level is None`.  Since `func_memory_level is None` currently means the result will  *never* be cached, it shouldn't have been wrapped at all.. and so clearly the user's confused about the design, and we can warn them.

I would be glad to submit a PR on this, but wanted to open things to discussion first.

### PR DESCRIPTION
**Pull request by AlexandreAbraham**
Fix #457

* In CacheMixin, make memory levels mandatory for function `_cache`. It is logic and they are never used without them.
* Fix logic in `cache` function (if called with a valid memory, no cache)
* Add tests
* Fix cache levels for several calls

### CODE REVIEW

File: nilearn/_utils/cache_mixin.py [hunk 1]
```diff
@@ -128,15 +128,14 @@ def cache(func, memory, func_memory_level=None, memory_level=None,
...
     verbose = kwargs.get('verbose', 0)
 
-    if (func_memory_level is None
-            or memory_level is None
-            or memory is None
-            or memory_level < func_memory_level):
-        memory = Memory(cachedir=None, verbose=verbose)
-    else:
+    if (func_memory_level is None and memory_level is not None or
+            memory_level is None and func_memory_level is not None):
```
Comment (bthirion) [inline nilearn/_utils/cache_mixin.py]: Could you add parentheses to help parsing ? I'm not sure I get ti right.

Comment (bthirion) [inline nilearn/_utils/cache_mixin.py]: And if this is not too complicated, explain the logic...

File: nilearn/input_data/nifti_masker.py [hunk 2]
```diff
@@ -186,7 +182,7 @@ def fit(self, imgs=None, y=None):
         # Resampling: allows the user to change the affine, the shape or both
         if self.verbose > 0:
             print("[%s.fit] Resampling mask" % self.__class__.__name__)
-        self.mask_img_ = self._cache(image.resample_img, func_memory_level=1)(
+        self.mask_img_ = self._cache(image.resample_img, 1)(
```
Comment (bthirion) [inline nilearn/input_data/nifti_masker.py]: In general, what is the strategy to set memory_level / func_memory elvel to 1 or 2 ? Should there be a pragraph on it in the doc ?

**Comment by bthirion** [comment]
Please notice a travis failure on Python 3.

File: nilearn/_utils/cache_mixin.py (hunk 1 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/cache_mixin.py]: Fixed.

File: nilearn/input_data/nifti_masker.py (hunk 2 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_masker.py]: Each function has a reference memory_level (func_memory_level). If the user memory_level if equal or greater to func_memory_level, then the function is cached. The goal is, for example, to have several layers of caching. In NiftiMasker:
* level 1 only cache the output (masked data)
* level 2 will cache intermediate step of the pretreatment.

File: nilearn/_utils/cache_mixin.py [hunk 3]
```diff
@@ -108,11 +108,11 @@ def cache(func, memory, func_memory_level=None, memory_level=None,
...
 
-    func_memory_level: int, optional
+    func_memory_level: int
         The memory_level from which caching must be enabled for the wrapped
         function.
 
-    memory_level: int, optional
+    memory_level: int
         The memory_level used to determine if function call must
         be cached or not (if user_memory_level is equal of grater than
```
Comment (lesteve) [inline nilearn/_utils/cache_mixin.py]: While you are at it: typo here grater -> gr**e**ater

File: nilearn/_utils/cache_mixin.py [hunk 4]
```diff
@@ -108,11 +108,11 @@ def cache(func, memory, func_memory_level=None, memory_level=None,
     memory: instance of joblib.Memory or string
         Used to cache the function call.
 
-    func_memory_level: int, optional
+    func_memory_level: int
```
Comment (lesteve) [inline nilearn/_utils/cache_mixin.py]: Looks like you meant to make both `func_memory_level` and `memory_level` parameters required but didn't do it in the function signature.

Also just wondering whether something more explicit than func_memory_level could be used ... No great suggestions other than reference_memory_level though.

File: nilearn/_utils/cache_mixin.py (hunk 3 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/cache_mixin.py]: Fixed.

File: nilearn/_utils/cache_mixin.py (hunk 4 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/cache_mixin.py]: Is there a standard way to precise that in the doc?

Also, I'm not fond of this name but it already changed twice...

File: nilearn/_utils/cache_mixin.py [hunk 5]
```diff
@@ -128,28 +128,33 @@ def cache(func, memory, func_memory_level=None, memory_level=None,
...
     verbose = kwargs.get('verbose', 0)
 
-    if (func_memory_level is None
-            or memory_level is None
-            or memory is None
-            or memory_level < func_memory_level):
-        memory = Memory(cachedir=None, verbose=verbose)
-    else:
+    # memory_level and func_memory_level must be both None or both integers.
+    none_params = (memory_level, func_memory_level).count(None)
```
Comment (lesteve) [inline nilearn/_utils/cache_mixin.py]: You are only checking half of your comment. Maybe something like that:

```python
memory_params = [memory_level, func_memory_level]
both_params_integers = all(isinstance(param, int) for param in memory_params)
both_params_none = all(param is None for param in memory_params)

if not (both_params_integers or both_params_none):
    raise ValueError(...)
```

File: nilearn/_utils/cache_mixin.py (hunk 4 above)
Comment (lesteve) [inline nilearn/_utils/cache_mixin.py]: Sorry I wasn't clear what I meant is that you didn't do what you announced in the description of the PR.

> In CacheMixin, make memory levels mandatory for function _cache. It is logic and they are never used without them.

Both `func_memory_level` and `memory_level` are still optional parameters atm ..

Comment (AlexandreAbraham) [inline nilearn/_utils/cache_mixin.py]: > In CacheMixin, make memory levels mandatory for function *_cache*. It is logic and they are never used without them.

I talk about function `_cache`, for which `func_memory_level` is now a mandatory parameter. `cache` is another function and can be called without them.

File: nilearn/_utils/cache_mixin.py (hunk 5 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/cache_mixin.py]: I like your version. Replaced!

File: nilearn/_utils/cache_mixin.py (hunk 4 above)
Comment (lesteve) [inline nilearn/_utils/cache_mixin.py]: Fair enough, but then func_memory_level is optional and the docstring should be consistent.

File: nilearn/_utils/cache_mixin.py [hunk 6]
```diff
@@ -108,13 +108,13 @@ def cache(func, memory, func_memory_level=None, memory_level=None,
     memory: instance of joblib.Memory or string
         Used to cache the function call.
 
-    func_memory_level: int, optional
+    func_memory_level: int
         The memory_level from which caching must be enabled for the wrapped
         function.
 
-    memory_level: int, optional
+    memory_level: int
```
Comment (lesteve) [inline nilearn/_utils/cache_mixin.py]: same thing here, optional shouldn't have been removed.

Comment (AlexandreAbraham) [inline nilearn/_utils/cache_mixin.py]: Fixed.

File: nilearn/input_data/nifti_labels_masker.py [hunk 7]
```diff
@@ -214,16 +214,14 @@ def transform(self, imgs, confounds=None):
                 self._resampled_labels_img_ = self.labels_img_
             if not _check_same_fov(imgs, self._resampled_labels_img_):
                 logger.log("resampling labels", verbose=self.verbose)
-                self._resampled_labels_img_ = self._cache(image.resample_img,
-                    func_memory_level=1)(
+                self._resampled_labels_img_ = self._cache(image.resample_img, 2)(
```
Comment (lesteve) [inline nilearn/input_data/nifti_labels_masker.py]: Any reason why you changed func_memory_level in this file? Used to be 1 now is 2.

Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_labels_masker.py]: Good point! The logic is that at level 1, only the signals should be cached. And at level 2, all intermediate steps should be cached. But the level 1 is missing here.

**Comment by AlexandreAbraham** [comment]
@lesteve Ready for next round ;)

**Comment by lesteve** [comment]
> @lesteve Ready for next round ;)

Not clear to me what the target_resampling=None has to do with this PR. Can you enlighten me ?

**Comment by AlexandreAbraham** [comment]
The code handling `target_resampling` was really messy and prevented me to do proper caching. I had to refactor it. I took this occasion to make it clearer.

**Comment by lesteve** [comment]
Looks like you need a rebase to fix the merge conflict.

File: nilearn/input_data/nifti_maps_masker.py [hunk 8]
```diff
@@ -149,28 +190,11 @@ def fit(self, X=None, y=None):
...
-                                          copy=True, order="C")
-            maps_affine = _utils.as_ndarray(
-                                            self.maps_img_.get_affine())
-            self.maps_img_ = new_img_like(self.maps_img_, maps_data,
-                                          maps_affine)
+            _assert_same_fov(mask=self.mask_img_, maps=self.maps_img_)
 
         elif self.resampling_target == "mask" and self.mask_img_ is not None:
-            logger.log("resampling regions", verbose=self.verbose)
+            if self.verbose > 0:
```
Comment (lesteve) [inline nilearn/input_data/nifti_maps_masker.py]: For better or worse it seems that logger is used consistently in this file. Not sure that this PR is the right place to replace logger.log by print statement.

File: nilearn/_utils/niimg_conversions.py [hunk 9]
```diff
@@ -33,6 +33,23 @@ def _check_same_fov(img1, img2):
             and np.allclose(img1.get_affine(), img2.get_affine()))
 
 
+def _assert_same_fov(**kwargs):
+    """ Assert the equivalence of all provided images. Parameter names are
+        used to generate user friendly error message.
+    """
+    errors = []
+    for (a_name, a_img), (b_name, b_img) in itertools.combinations(
```
Comment (lesteve) [inline nilearn/_utils/niimg_conversions.py]: The use of itertools.combinations looks quite smart but you could just take the first image and make sure it has the same field of view as the remaining ones, couldn't you ?

File: nilearn/_utils/niimg_conversions.py [hunk 10]
```diff
@@ -33,6 +33,23 @@ def _check_same_fov(img1, img2):
             and np.allclose(img1.get_affine(), img2.get_affine()))
 
 
+def _assert_same_fov(**kwargs):
```
Comment (lesteve) [inline nilearn/_utils/niimg_conversions.py]: It is a bit of a shame that _assert_same_fov doesn't reuse _check_same_fov given their similar functionality.

File: nilearn/input_data/nifti_labels_masker.py [hunk 11]
```diff
@@ -205,51 +251,34 @@ def transform(self, imgs, confounds=None):
...
-            logger.log("resampling images", verbose=self.verbose)
-            imgs = self._cache(image.resample_img, func_memory_level=1)(
-                imgs, interpolation="continuous",
-                target_shape=self.labels_img_.shape,
-                target_affine=self.labels_img_.get_affine())
-        else:
-            self._resampled_labels_img_ = self.labels_img_
+                        target_affine=imgs.get_affine())
+
+        region_signals = self._cache(_extract_signals, 1, ignore=['verbose'])(
```
Comment (lesteve) [inline nilearn/input_data/nifti_labels_masker.py]: Don't you need to add more ignored parameters, for example memory and memory_level?

File: nilearn/input_data/nifti_labels_masker.py [hunk 12]
```diff
@@ -18,6 +19,51 @@
 from .. import image
 
 
+def _extract_signals(imgs, labels_img, background_label, smoothing_fwhm,
+                     t_r, standardize, detrend, low_pass, high_pass,
+                     confounds, memory, memory_level,
+                     resample_on_labels=False, verbose=0):
+    if verbose > 0:
```
Comment (lesteve) [inline nilearn/input_data/nifti_labels_masker.py]: Maybe a short docstring would be nice

File: nilearn/input_data/nifti_labels_masker.py [hunk 13]
```diff
@@ -205,51 +251,34 @@ def transform(self, imgs, confounds=None):
         """
         self._check_fitted()
 
-        logger.log("loading images: %s" %
-                   _utils._repr_niimgs(imgs)[:200], verbose=self.verbose)
-        imgs = _utils.check_niimg_4d(imgs)
+        # We handle the resampling of labels separately because the affine of
+        # the labels image should not impact the extraction of the signal.
 
+        if not hasattr(self, '_resampled_labels_img_'):
```
Comment (lesteve) [inline nilearn/input_data/nifti_labels_masker.py]: Just curious, where is _resampled_labels_img_ likely to be set prior to this statement?

File: nilearn/input_data/nifti_maps_masker.py [hunk 14]
```diff
@@ -2,22 +2,62 @@
...
-from .._utils.niimg_conversions import _check_same_fov
-from .._utils import _compose_err_msg
+from .._utils.cache_mixin import cache
+from .._utils.niimg_conversions import _check_same_fov, _assert_same_fov
 from .. import signal
 from .. import region
 from .. import image
 
 
+def _extract_signals(imgs, maps_img, smoothing_fwhm,
```
Comment (lesteve) [inline nilearn/input_data/nifti_maps_masker.py]: Seems very similar to _extract_signals in nifti_labels_masker.py. Maybe worth looking at some refactoring here.

File: nilearn/input_data/nifti_maps_masker.py [hunk 15]
```diff
@@ -221,75 +246,64 @@ def transform(self, imgs, confounds=None):
...
+                    not _check_same_fov(ref_img, self.mask_img_)):
+                if self.verbose > 0:
+                    print("Resampling mask")
+                self._resampled_mask_img_ = self._cache(image.resample_img, 1)(
+                        self.mask_img_, interpolation="nearest",
+                        target_shape=ref_img.shape[:3],
+                        target_affine=ref_img.get_affine())
+
+        region_signals, labels_ = self._cache(
+            _extract_signals, 1, ignore=['verbose'])(
```
Comment (lesteve) [inline nilearn/input_data/nifti_maps_masker.py]: Do you need to ignore memory and memory_level here?

File: nilearn/tests/test_cache_mixin.py [hunk 16]
```diff
@@ -76,11 +76,14 @@ def test__safe_cache_flush():
...
-    job_glob = os.path.join(temp_dir, 'joblib', '*')
+    job_glob = os.path.join(temp_dir, 'joblib', 'nilearn', 'tests',
+                            'test_cache_mixin', 'f', '*')
     mem = Memory(cachedir=temp_dir, verbose=0)
-    cache_mixin.cache(f, mem)(2)
-    assert_true(len(glob.glob(job_glob)) == 0)
     cache_mixin.cache(f, mem, func_memory_level=2, memory_level=1)(2)
     assert_true(len(glob.glob(job_glob)) == 0)
+    cache_mixin.cache(f, Memory(cachedir=None))(2)
+    assert_true(len(glob.glob(job_glob)) == 0)
```
Comment (lesteve) [inline nilearn/tests/test_cache_mixin.py]: `assert_equal(a, b)` gives slightly nicer error messages than `assert_true(a == b)`.

File: nilearn/input_data/nifti_labels_masker.py (hunk 13 above)
Comment (lesteve) [inline nilearn/input_data/nifti_labels_masker.py]: Looking at it a bit closer, maybe from multiple calls to transform ?

File: nilearn/input_data/nifti_maps_masker.py (hunk 8 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_maps_masker.py]: This is a PR to fix bugs so I took the occasion to fix other problems. I can remove it if it's necessary though.

File: nilearn/_utils/niimg_conversions.py (hunk 9 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg_conversions.py]: It's possible but I wanted to have the best error message possible. Note that this function is called with one or two params.

File: nilearn/_utils/niimg_conversions.py (hunk 10 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg_conversions.py]: They can be merged if I add a parameter to raise an error.

File: nilearn/input_data/nifti_labels_masker.py (hunk 11 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_labels_masker.py]: Fixed.

File: nilearn/input_data/nifti_labels_masker.py (hunk 12 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_labels_masker.py]: Fixed.

File: nilearn/input_data/nifti_labels_masker.py (hunk 13 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_labels_masker.py]: Yes. It is a caching system for multiple calls.

File: nilearn/input_data/nifti_maps_masker.py (hunk 15 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_maps_masker.py]: Fixed

File: nilearn/tests/test_cache_mixin.py (hunk 16 above)
Comment (AlexandreAbraham) [inline nilearn/tests/test_cache_mixin.py]: I'm afraid to fix stuff since people tell me that it should be part of another PR :P.

File: nilearn/input_data/nifti_maps_masker.py (hunk 8 above)
Comment (lesteve) [inline nilearn/input_data/nifti_maps_masker.py]: Can you remind me what was the problem with the logger?

Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_maps_masker.py]: Nothing but the fact that it is not consistant with the rest of the API. I know that we once talked about removing it. Gaël has also suggested another logging system to replace it.

**Comment by lesteve** [comment]
FYI this PR needs a rebase.

File: nilearn/_utils/niimg_conversions.py (hunk 10 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg_conversions.py]: Merged.

File: nilearn/tests/test_cache_mixin.py (hunk 16 above)
Comment (AlexandreAbraham) [inline nilearn/tests/test_cache_mixin.py]: Fixed.

File: nilearn/input_data/nifti_maps_masker.py (hunk 14 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_maps_masker.py]: Yes, but this is not the point of this PR. This is not as trivial as it seems. In particular, this is not the only code that can be factorized. It can be merged with the code coming from NiftiSpheresMasker and Searchlight (which have very similar code too). I was thinking of a BaseElementMasker that iterates on some kind of ROI iterator (list of ROI or labels or maps). But this require a lot of work. This is the step after!

File: nilearn/_utils/niimg_conversions.py [hunk 17]
```diff
@@ -24,14 +24,25 @@ def _check_fov(img, affine, shape):
             np.allclose(img.get_affine(), affine))
 
 
-def _check_same_fov(img1, img2):
-    """ Return True if img1 and img2 have the same field of view
-        (shape and affine), False elsewhere.
+def _check_same_fov(*args, **kwargs):
+    """ Check the equivalence of all provided images. Parameter names are
+        used to generate user friendly error message.
```
Comment (GaelVaroquaux) [inline nilearn/_utils/niimg_conversions.py]: That function needs a better docstring. First, the notion of "equivalence" of images should be made explicit (am I right in understanding that its: same shape and affine?). Second, what the kwargs are should be detailed.

It doesn't have to be of the quality of user-facing docs, it's more a question of devs easily understanding what the function does.

File: nilearn/input_data/multi_nifti_masker.py [hunk 18]
```diff
@@ -178,9 +178,7 @@ def fit(self, imgs=None, y=None):
                 raise ValueError("Unknown value of mask_strategy '%s'. "
                     "Acceptable values are 'background' and 'epi'.")
 
-            self.mask_img_ = self._cache(
-                        compute_mask,
-                        func_memory_level=1,
+            self.mask_img_ = self._cache(compute_mask, 1,
```
Comment (GaelVaroquaux) [inline nilearn/input_data/multi_nifti_masker.py]: I would prefer, for readability, still having the name of the keyword argument in the code. It serves to explain what the '1' means.

File: nilearn/input_data/multi_nifti_masker.py [hunk 19]
```diff
@@ -201,8 +199,7 @@ def fit(self, imgs=None, y=None):
         # Resampling: allows the user to change the affine, the shape or both.
         if self.verbose > 0:
             print("[%s.transform] Resampling mask" % self.__class__.__name__)
-        self.mask_img_ = self._cache(image.resample_img,
-                                    func_memory_level=1)(
+        self.mask_img_ = self._cache(image.resample_img, 1)(
```
Comment (GaelVaroquaux) [inline nilearn/input_data/multi_nifti_masker.py]: Same remark about keeping the name of the kwarg.

File: nilearn/input_data/nifti_labels_masker.py [hunk 20]
```diff
@@ -205,51 +253,35 @@ def transform(self, imgs, confounds=None):
...
-            imgs = self._cache(image.resample_img, func_memory_level=1)(
-                imgs, interpolation="continuous",
-                target_shape=self.labels_img_.shape,
-                target_affine=self.labels_img_.get_affine())
-        else:
-            self._resampled_labels_img_ = self.labels_img_
+                        target_affine=imgs.get_affine())
+
+        region_signals = self._cache(
+            _extract_signals, 1, ignore=['verbose', 'memory', 'memory_level'])(
```
Comment (GaelVaroquaux) [inline nilearn/input_data/nifti_labels_masker.py]: I'd be happier with the kwarg being explicit :)

File: nilearn/input_data/nifti_labels_masker.py [hunk 21]
```diff
@@ -205,51 +253,35 @@ def transform(self, imgs, confounds=None):
...
+            # Images
+            imgs, self._resampled_labels_img_, self.background_label,
+            # Pre-treatments
```
Comment (GaelVaroquaux) [inline nilearn/input_data/nifti_labels_masker.py]: I think that in English, this is know as 'pre-processing". "Treatement" being a bad translation.

File: nilearn/_utils/cache_mixin.py [hunk 22]
```diff
@@ -165,7 +172,7 @@ class CacheMixin(object):
     parameter to self._cache(). See _cache() documentation for details.
     """
 
-    def _cache(self, func, func_memory_level=1, **kwargs):
+    def _cache(self, func, func_memory_level, **kwargs):
```
Comment (GaelVaroquaux) [inline nilearn/_utils/cache_mixin.py]: Is there a specific reason that you removed the default value?

Comment (AlexandreAbraham) [inline nilearn/_utils/cache_mixin.py]: I just wanted to make it explicit and mandatory. No strong feelings about that.

Comment (GaelVaroquaux) [inline nilearn/_utils/cache_mixin.py]: > I just wanted to make it explicit and mandatory. No strong feelings about that.

I'd rather not bother people if they are not thinking about this kind of
problem: it's a valid usecase to want to cache without controling the
granularity.

File: nilearn/_utils/niimg_conversions.py (hunk 17 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg_conversions.py]: Fixed.

File: nilearn/input_data/multi_nifti_masker.py (hunk 18 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/multi_nifti_masker.py]: Fixed.

File: nilearn/input_data/multi_nifti_masker.py (hunk 19 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/multi_nifti_masker.py]: Fixed.

File: nilearn/input_data/nifti_labels_masker.py (hunk 20 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_labels_masker.py]: Fixed.

File: nilearn/input_data/nifti_labels_masker.py (hunk 21 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_labels_masker.py]: Fixed.

**Comment by GaelVaroquaux** [comment]
OK. Thx. Merging.
